### PR TITLE
First and Last Aggregator

### PR DESCRIPTION
**Pull request by acslk**
This PR implements the 'first' and 'last' aggregator discussed in #2845.

The first and last aggregator can be used in the following format

```
{ "type" : "doubleFirst", "name" : <output_name>, "fieldName" : <metric_name>}
{ "type" : "longFirst", "name" : <output_name>, "fieldName" : <metric_name>}
{ "type" : "doubleLast", "name" : <output_name>, "fieldName" : <metric_name>}
{ "type" : "longLast", "name" : <output_name>, "fieldName" : <metric_name>}
```

The first aggregator output the value of _fieldName_ with the smallest timestamp (using the __time column), while last aggregator output the value of _fieldName_ with the largest timestamp. In case of multiple first and last times, one of them will be selected arbitrarily.

### CODE REVIEW

**Comment by fjy** [comment]
@acslk we need docs in docs/content/querying aggregations doc

File: common/src/main/java/io/druid/collections/SerializablePair.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+ */
+
+package io.druid.collections;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.metamx.common.Pair;
+
+public class SerializablePair<T1, T2> extends Pair<T1, T2>
+{
+  public SerializablePair(@JsonProperty T1 lhs, @JsonProperty T2 rhs)
```
Comment (fjy) [inline common/src/main/java/io/druid/collections/SerializablePair.java]: do you need a @JsonCreator annotation?

File: processing/src/test/java/io/druid/query/groupby/GroupByQueryRunnerTest.java [hunk 2]
```diff
@@ -1333,6 +1335,42 @@ public void testGroupByWithCardinality()
   }
 
   @Test
+  public void testGroupByWithFirstLast()
```
Comment (fjy) [inline processing/src/test/java/io/druid/query/groupby/GroupByQueryRunnerTest.java]: can we add a test where we select a non valid time range?

Comment (fjy) [inline processing/src/test/java/io/druid/query/groupby/GroupByQueryRunnerTest.java]: or rather, a time range with no results

Comment (fjy) [inline processing/src/test/java/io/druid/query/groupby/GroupByQueryRunnerTest.java]: also for timeseries and topN

File: common/src/main/java/io/druid/collections/SerializablePair.java (hunk 1 above)
Comment (acslk) [inline common/src/main/java/io/druid/collections/SerializablePair.java]: I don't need it for this PR since I create the SerializablePair using object map in AggregatorFactory instead of using the object mapper, but it should be useful if this class is to be used in the future.

File: processing/src/main/java/io/druid/query/aggregation/FirstAggregatorFactory.java [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+      return new DoubleFirstBufferAggregator(
+          metricFactory.makeLongColumnSelector(Column.TIME_COLUMN_NAME),
+          metricFactory.makeFloatColumnSelector(fieldName)
+      );
+    }
+    throw new IAE("undefined type");
+  }
+
+  @Override
+  public Comparator getComparator()
```
Comment (gianm) [inline processing/src/main/java/io/druid/query/aggregation/FirstAggregatorFactory.java]: The comparator from `getComparator` is used for ranking by metric in topNs and groupBys; we probably would rather rank by the actual last value rather than by the timestamp.

File: common/src/main/java/io/druid/collections/SerializablePair.java (hunk 1 above)
Comment (gianm) [inline common/src/main/java/io/druid/collections/SerializablePair.java]: It's good code cleanliness to have deserializers when we have serializers, so let's add both (and a test for both).

**Comment by gianm** [comment]
These should be documented in aggregations.md.

File: processing/src/main/java/io/druid/query/aggregation/LastAggregatorFactory.java [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+
+  private final String fieldName;
+  private final String name;
+  private final String value;
+
+  @JsonCreator
+  public LastAggregatorFactory(
+      @JsonProperty("name") String name,
+      @JsonProperty("fieldName") final String fieldName,
+      @JsonProperty("value") String value
```
Comment (gianm) [inline processing/src/main/java/io/druid/query/aggregation/LastAggregatorFactory.java]: `valueType` is probably clearer

File: processing/src/main/java/io/druid/query/aggregation/LastAggregatorFactory.java [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+
+public class LastAggregatorFactory extends AggregatorFactory
+{
+  private static final byte CACHE_TYPE_ID = 0x11;
+  private static final String TYPE_LONG = "long";
+  private static final String TYPE_DOUBLE = "double";
+
+  private final String fieldName;
+  private final String name;
+  private final String value;
```
Comment (gianm) [inline processing/src/main/java/io/druid/query/aggregation/LastAggregatorFactory.java]: This could/should be an enum (possibly the already-existing `ValueType`)

File: processing/src/main/java/io/druid/query/aggregation/LastAggregatorFactory.java [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+
+  @JsonCreator
+  public LastAggregatorFactory(
```
Comment (gianm) [inline processing/src/main/java/io/druid/query/aggregation/LastAggregatorFactory.java]: Hmm, maybe `longLast` and `doubleLast` aren't so bad… it looks like there is really little code shared between them.

File: processing/src/main/java/io/druid/query/aggregation/LastAggregatorFactory.java [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+  }
+
+  @Override
+  public Object combine(Object lhs, Object rhs)
+  {
+    return (((SerializablePair<Long, Object>) lhs).lhs > ((SerializablePair<Long, Object>) rhs).lhs) ? lhs : rhs;
+  }
+
+  @Override
+  public AggregatorFactory getCombiningFactory()
```
Comment (gianm) [inline processing/src/main/java/io/druid/query/aggregation/LastAggregatorFactory.java]: Hmm, this is a bit weird, but I guess you're trying to avoid having both a LongLastAggregator and LongLastCombiningAggregator. I can't think of a better way to structure that right now and I'm not sure which of those two options I like better…

File: processing/src/main/java/io/druid/query/aggregation/LastAggregatorFactory.java [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+  {
+    return Arrays.asList(Column.TIME_COLUMN_NAME, fieldName);
+  }
+
+  @Override
+  public byte[] getCacheKey()
+  {
+    byte[] fieldNameBytes = StringUtils.toUtf8(fieldName);
+
+    return ByteBuffer.allocate(1 + fieldNameBytes.length).put(CACHE_TYPE_ID).put(fieldNameBytes).array();
```
Comment (gianm) [inline processing/src/main/java/io/druid/query/aggregation/LastAggregatorFactory.java]: `value` should be part of this too, as it can change the results.

File: processing/src/main/java/io/druid/query/aggregation/LastAggregatorFactory.java [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+    if (value.equals(TYPE_LONG)) {
+      return Longs.BYTES * 2;
+    } else if (value.equals(TYPE_DOUBLE)) {
+      return Longs.BYTES + Doubles.BYTES;
+    }
+    throw new IAE("undefined type");
+  }
+
+  @Override
+  public Object getAggregatorStartValue()
```
Comment (gianm) [inline processing/src/main/java/io/druid/query/aggregation/LastAggregatorFactory.java]: It's ok to throw UnsupportedOperationException here, nothing calls this and we should probably get rid of it at some point.

File: processing/src/main/java/io/druid/query/aggregation/LastAggregatorFactory.java [hunk 10]
```diff
@@ -0,0 +1,335 @@
...
+      return new DoubleLastBufferAggregator(
+          metricFactory.makeLongColumnSelector(Column.TIME_COLUMN_NAME),
+          metricFactory.makeFloatColumnSelector(fieldName)
+      );
+    }
+    throw new IAE("undefined type");
+  }
+
+  @Override
+  public Comparator getComparator()
```
Comment (gianm) [inline processing/src/main/java/io/druid/query/aggregation/LastAggregatorFactory.java]: The comparator from `getComparator` is used for ranking by metric in topNs and groupBys; we probably would rather rank by the actual last value rather than by the timestamp.

File: processing/src/main/java/io/druid/query/aggregation/FirstAggregatorFactory.java [hunk 11]
```diff
@@ -0,0 +1,338 @@
...
+import io.druid.segment.ObjectColumnSelector;
+import io.druid.segment.column.Column;
+
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+
+public class FirstAggregatorFactory extends AggregatorFactory
```
Comment (gianm) [inline processing/src/main/java/io/druid/query/aggregation/FirstAggregatorFactory.java]: Generally similar comments to LastAggregatorFactory

File: processing/src/test/java/io/druid/query/groupby/GroupByQueryRunnerTest.java (hunk 2 above)
Comment (gianm) [inline processing/src/test/java/io/druid/query/groupby/GroupByQueryRunnerTest.java]: Please also include a test that this works at ingestion time – ie in an incremental index. (some of the use cases from the user lists involved wanting that behavior)

**Comment by gianm** [comment]
I think this won't work at indexing time as-is; we would need a serde for writing out columns that have the (timestamp, last value) pairs in them.

**Comment by acslk** [comment]
@gianm Using first/last aggregator at ingestion time is kind of tricky with the value we want to store. Ideally, we want to persist the first and last metric as long/double column so other aggregator such as sum can aggregate them. However, doing so would cause merging persisted data to be incorrect since no time value are stored for the metric. If the values are instead stored as time value pair, the column could not be aggregated by standard aggregators. Basically the problem is that we want an intermediate storage format for merging and a different final storage format for querying, and this cannot be done in the ingestion process. Now I’ll just leave the comment in aggregations.md that first/last aggregators can't be used at ingestion time.

**Comment by acslk** [comment]
It seems that there are very little shared code between long and double type aggregator, so I changed the syntax to be the same with max, min, and sum with doubleFirst/Last and longFirst/Last in the newer commit.

File: processing/src/main/java/io/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+
+public class DoubleFirstAggregatorFactory extends AggregatorFactory
+{
+  private static final byte CACHE_TYPE_ID = 16;
```
Comment (fjy) [inline processing/src/main/java/io/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java]: we should really define these constants in another file somewhere. It is getting more and more difficult to track available values

Comment (drcrallen) [inline processing/src/main/java/io/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java]: Or some other better way to track cache unique key guarantees. Right now it is pretty much impossible for extensions to guarantee uniqueness of id across them.

Comment (acslk) [inline processing/src/main/java/io/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java]: we can store the cache keys in a helper class similar to how DimFilter cache keys are stored, but I think that can go in another PR

**Comment by fjy** [comment]
👍

File: processing/src/main/java/io/druid/query/aggregation/last/DoubleLastAggregatorFactory.java [hunk 13]
```diff
@@ -0,0 +1,250 @@
...
+  {
+    byte[] fieldNameBytes = StringUtils.toUtf8(fieldName);
+
+    return ByteBuffer.allocate(1 + fieldNameBytes.length).put(CACHE_TYPE_ID).put(fieldNameBytes).array();
+  }
+
+  @Override
+  public String getTypeName()
+  {
+    return "float";
```
Comment (navis) [inline processing/src/main/java/io/druid/query/aggregation/last/DoubleLastAggregatorFactory.java]: I understand it's not useable for ingestion time, but at least shouldn't we return float or long here?

Comment (acslk) [inline processing/src/main/java/io/druid/query/aggregation/last/DoubleLastAggregatorFactory.java]: Sorry if I misunderstood, but are you suggesting to have getTypeName return some type other than float or long?

Comment (navis) [inline processing/src/main/java/io/druid/query/aggregation/last/DoubleLastAggregatorFactory.java]: Ah, forget that. Now I understand the intention.

File: processing/src/main/java/io/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java (hunk 12 above)
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java]: fixing in a larger way is outside the scope of this PR

File: processing/src/main/java/io/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java [hunk 14]
```diff
@@ -0,0 +1,249 @@
...
+    return new DoubleFirstBufferAggregator(
+        metricFactory.makeLongColumnSelector(Column.TIME_COLUMN_NAME),
+        metricFactory.makeFloatColumnSelector(fieldName)
+    );
+  }
+
+  @Override
+  public Comparator getComparator()
+  {
+    return new Comparator()
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java]: This can be a static object

File: processing/src/main/java/io/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java [hunk 15]
```diff
@@ -0,0 +1,249 @@
...
+      {
+        return Doubles.compare(((SerializablePair<Long, Double>) o1).rhs, ((SerializablePair<Long, Double>) o2).rhs);
+      }
+    };
+  }
+
+  @Override
+  public Object combine(Object lhs, Object rhs)
+  {
+    return (((SerializablePair<Long, Double>) lhs).lhs <= ((SerializablePair<Long, Double>) rhs).lhs) ? lhs : rhs;
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java]: And this should be able to use the static comparator to be a little less full of `(((())))(`

File: processing/src/main/java/io/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java [hunk 16]
```diff
@@ -0,0 +1,249 @@
...
+  @Override
+  public List<String> requiredFields()
+  {
+    return Arrays.asList(Column.TIME_COLUMN_NAME, fieldName);
+  }
+
+  @Override
+  public byte[] getCacheKey()
+  {
+    byte[] fieldNameBytes = StringUtils.toUtf8(fieldName);
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java]: Suggest terminating with `0xFF` (can be added in byte buffer)

File: processing/src/main/java/io/druid/query/aggregation/first/LongFirstAggregatorFactory.java [hunk 17]
```diff
@@ -0,0 +1,248 @@
...
+    return new LongFirstBufferAggregator(
+        metricFactory.makeLongColumnSelector(Column.TIME_COLUMN_NAME),
+        metricFactory.makeLongColumnSelector(fieldName)
+    );
+  }
+
+  @Override
+  public Comparator getComparator()
+  {
+    return new Comparator()
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/aggregation/first/LongFirstAggregatorFactory.java]: same suggestion about static comparator

File: processing/src/main/java/io/druid/query/aggregation/first/LongFirstAggregatorFactory.java [hunk 18]
```diff
@@ -0,0 +1,248 @@
...
+      {
+        return Longs.compare(((SerializablePair<Long, Long>) o1).rhs, ((SerializablePair<Long, Long>) o2).rhs);
+      }
+    };
+  }
+
+  @Override
+  public Object combine(Object lhs, Object rhs)
+  {
+    return (((SerializablePair<Long, Long>) lhs).lhs <= ((SerializablePair<Long, Long>) rhs).lhs) ? lhs : rhs;
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/aggregation/first/LongFirstAggregatorFactory.java]: And using the static comparator here

File: processing/src/main/java/io/druid/query/aggregation/first/LongFirstAggregatorFactory.java [hunk 19]
```diff
@@ -0,0 +1,248 @@
...
+  @Override
+  public List<String> requiredFields()
+  {
+    return Arrays.asList(Column.TIME_COLUMN_NAME, fieldName);
+  }
+
+  @Override
+  public byte[] getCacheKey()
+  {
+    byte[] fieldNameBytes = StringUtils.toUtf8(fieldName);
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/aggregation/first/LongFirstAggregatorFactory.java]: suggest terminating in `0xFF`

File: processing/src/main/java/io/druid/query/aggregation/last/DoubleLastAggregatorFactory.java [hunk 20]
```diff
@@ -0,0 +1,250 @@
...
+    return new DoubleLastBufferAggregator(
+        metricFactory.makeLongColumnSelector(Column.TIME_COLUMN_NAME),
+        metricFactory.makeFloatColumnSelector(fieldName)
+    );
+  }
+
+  @Override
+  public Comparator getComparator()
+  {
+    return new Comparator()
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/aggregation/last/DoubleLastAggregatorFactory.java]: static comparator again

File: processing/src/main/java/io/druid/query/aggregation/last/DoubleLastAggregatorFactory.java [hunk 21]
```diff
@@ -0,0 +1,250 @@
...
+      {
+        return Doubles.compare(((SerializablePair<Long, Double>) o1).rhs, ((SerializablePair<Long, Double>) o2).rhs);
+      }
+    };
+  }
+
+  @Override
+  public Object combine(Object lhs, Object rhs)
+  {
+    return (((SerializablePair<Long, Double>) lhs).lhs > ((SerializablePair<Long, Double>) rhs).lhs) ? lhs : rhs;
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/aggregation/last/DoubleLastAggregatorFactory.java]: and using comparator here

File: processing/src/main/java/io/druid/query/aggregation/last/DoubleLastAggregatorFactory.java [hunk 22]
```diff
@@ -0,0 +1,250 @@
...
+  @Override
+  public List<String> requiredFields()
+  {
+    return Arrays.asList(Column.TIME_COLUMN_NAME, fieldName);
+  }
+
+  @Override
+  public byte[] getCacheKey()
+  {
+    byte[] fieldNameBytes = StringUtils.toUtf8(fieldName);
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/aggregation/last/DoubleLastAggregatorFactory.java]: same suggestion on terminating with `0xFF`

File: processing/src/main/java/io/druid/query/aggregation/last/LongLastAggregatorFactory.java [hunk 23]
```diff
@@ -0,0 +1,249 @@
...
+  @Override
+  public List<String> requiredFields()
+  {
+    return Arrays.asList(Column.TIME_COLUMN_NAME, fieldName);
+  }
+
+  @Override
+  public byte[] getCacheKey()
+  {
+    byte[] fieldNameBytes = StringUtils.toUtf8(fieldName);
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/aggregation/last/LongLastAggregatorFactory.java]: same suggestion on `0xFF`

**Comment by drcrallen** [comment]
Cool, relatively minor comments over all. I have to admin, my brain started to glaze over about 1/3 of the way through, so I'm hoping there's nothing too crazy hidden in the java-primitives-require-code-copying madness.

File: processing/src/main/java/io/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java (hunk 15 above)
Comment (acslk) [inline processing/src/main/java/io/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java]: Combine uses a different compare function than the Comparator above, as combine compares time on the lhs while getComparator compares the value on the rhs. It's probably still a good idea to put the combine comparator as static object since (...lhs).lhs seems pretty confusing

File: processing/src/main/java/io/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java (hunk 16 above)
Comment (acslk) [inline processing/src/main/java/io/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java]: what would be the intention of the terminating byte?

Comment (drcrallen) [inline processing/src/main/java/io/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java]: add a `0xFF` after this byte array when putting it in the byte buffer to ensure proper string termination.

File: processing/src/main/java/io/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java (hunk 14 above)
Comment (acslk) [inline processing/src/main/java/io/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java]: changed getComparator to return static object for all first/last aggregators

File: processing/src/main/java/io/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java (hunk 16 above)
Comment (acslk) [inline processing/src/main/java/io/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java]: added 0xFF byte to cache key for all first/last aggregators

File: processing/src/main/java/io/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java (hunk 15 above)
Comment (acslk) [inline processing/src/main/java/io/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java]: added a static time comparator for combine for all first/last aggregators

**Comment by drcrallen** [comment]
```
Tests in error: 
  GroupByQueryRunnerTest.testSubqueryWithFirstLast:5144 » Parse Unknown type[cla...
  GroupByQueryRunnerTest.testSubqueryWithFirstLast:5144 » Parse Unknown type[cla...
  GroupByQueryRunnerTest.testSubqueryWithFirstLast:5144 » Parse Unknown type[cla...
  GroupByQueryRunnerTest.testSubqueryWithFirstLast:5144 » Parse Unknown type[cla...
  GroupByQueryRunnerTest.testSubqueryWithFirstLast:5144 » Parse Unknown type[cla...
  GroupByQueryRunnerTest.testSubqueryWithFirstLast:5144 » Parse Unknown type[cla...
  GroupByQueryRunnerTest.testSubqueryWithFirstLast:5144 » Parse Unknown type[cla...
  GroupByQueryRunnerTest.testSubqueryWithFirstLast:5144 » Parse Unknown type[cla...
  GroupByQueryRunnerTest.testSubqueryWithFirstLast:5144 » Parse Unknown type[cla...
  GroupByQueryRunnerTest.testSubqueryWithFirstLast:5144 » Parse Unknown type[cla...
  GroupByQueryRunnerTest.testSubqueryWithFirstLast:5144 » Parse Unknown type[cla...
  GroupByQueryRunnerTest.testSubqueryWithFirstLast:5144 » Parse Unknown type[cla...
  GroupByQueryRunnerTest.testSubqueryWithFirstLast:5144 » Parse Unknown type[cla...
  GroupByQueryRunnerTest.testSubqueryWithFirstLast:5144 » Parse Unknown type[cla...
  GroupByQueryRunnerTest.testSubqueryWithFirstLast:5144 » Parse Unknown type[cla...

```

looks like real error

**Comment by acslk** [comment]
seems like incompatibility with the v2 nested groupBy change

**Comment by acslk** [comment]
The error in nested groupBy test is caused by outer query expecting data source to have long value for the metric column created by inner query's longFirst aggregator, but instead got the intermediate value type which is pair<time, value> since inner query result isn't finalized. I feel like inner query's result should be finalized by the definition of query as datasource, but doing so would break aggregators in outer query that depend on intermediate value such as HyperUnique. The fix I made parsed pair<time, long/double> as long/double in mapBasedRow.

**Comment by gianm** [comment]
Hmm, finalizing inner queries makes sense to me…

Is there any reason why they aren't? does anyone know?

**Comment by acslk** [comment]
Tried finalizing inner query result, the failing tests are testSubqueryWithHyperUniques for v1 and v2 and testSubqueryWithFirstLast (introduced in this PR) for v1. HyperUniques test failed because the outer query has a HyperUniques aggregator, which require the pre-finalized value from the inner query result. FirstLast test failed for v1 because v1 builds inner incremental index using combining factory, which assumes intermediate data format.

**Comment by gianm** [comment]
Hmm, does the "finalize" context param do anything on the inner query? IMO it makes sense for the inner query's value of "finalize" to be overridden. false makes more sense for a default, IMO, but it could be true if it needs to be for compat reasons. Either way, a user should also be able to set it to whatever the non-default value is.

Setting "finalize" appropriately should let the tests pass…

**Comment by acslk** [comment]
Currently "finalize" param for inner query is not used as inner query is run using mergeResults which rewrites query to set "finalize" to false. Finalizing inner query can be added by using mapped sequence on top of the inner query result, maybe in another PR?

**Comment by gianm** [comment]
@acslk is a finalization option not needed for making the first/last aggregator work?

**Comment by acslk** [comment]
The previous commit added parsing for pair for RowBasedMap which is used by nested groupBy v2, while v1 uses value from IncrementalIndexAdapter which uses aggregator's getLong/getFloat, so both versions works right now without finalization.

**Comment by gianm** [comment]
Ahh, I hadn't seen that before. hmm, that Pair parsing is kind of weird, though. Namely: What does it mean to "parse" a Pair as a number? Why is the rhs the special one? Why not treat it like a vector and take its magnitude? :)

I recognize the answer is: because that's what makes the first/last aggregators work. But that answer is somewhat unsatisfying – a finalization option would be a lot nicer.

**Comment by gianm** [comment]
hmm, I agree finalize = true makes the most sense, but I think in the case compatibility concerns win. So let's make it false by default.

File: processing/src/main/java/io/druid/query/aggregation/last/DoubleLastAggregatorFactory.java [hunk 24]
```diff
@@ -0,0 +1,243 @@
...
+  {
+    int result = name.hashCode();
+    result = 31 * result + fieldName.hashCode();
+    return result;
+  }
+
+  @Override
+  public String toString()
+  {
+    return "LongFirstAggregatorFactory{" +
```
Comment (jon-wei) [inline processing/src/main/java/io/druid/query/aggregation/last/DoubleLastAggregatorFactory.java]: should be "DoubleLastAggregatorFactory{"

File: processing/src/main/java/io/druid/query/aggregation/last/LongLastAggregatorFactory.java [hunk 25]
```diff
@@ -0,0 +1,243 @@
...
+  {
+    int result = name.hashCode();
+    result = 31 * result + fieldName.hashCode();
+    return result;
+  }
+
+  @Override
+  public String toString()
+  {
+    return "DoubleLastAggregatorFactory{" +
```
Comment (jon-wei) [inline processing/src/main/java/io/druid/query/aggregation/last/LongLastAggregatorFactory.java]: should be "LongLastAggregatorFactory{"

File: processing/src/main/java/io/druid/query/aggregation/first/LongFirstAggregatorFactory.java [hunk 26]
```diff
@@ -0,0 +1,250 @@
...
+
+    return ByteBuffer.allocate(2 + fieldNameBytes.length).put(CACHE_TYPE_ID).put(fieldNameBytes).put((byte)0xff).array();
```
Comment (jon-wei) [inline processing/src/main/java/io/druid/query/aggregation/first/LongFirstAggregatorFactory.java]: nit: can you put the chained function calls on separate lines? I think it makes it easier to see the composition of the cache key that way

**Comment by jon-wei** [comment]
some minor comments, looks good so far, will review again after query finalization comments from @gianm are resolved

**Comment by acslk** [comment]
Added option to finalize inner query, and also slightly changed how v1 build inner incrementalIndex. v1 strategy process inner query result by building IncrementalIndex on query result with aggregators from AggregatorFactory.getRequiredColumn(). When building IncrementalIndex, getCombiningFactory is called for the passed in aggregators. This makes sense for the merging runners that uses IncrementalIndex, but not so much for copying value from results. I parameterized whether or not to use combining factory so indexing inner query does not use combining factory.

File: processing/src/main/java/io/druid/query/aggregation/last/DoubleLastAggregatorFactory.java (hunk 24 above)
Comment (acslk) [inline processing/src/main/java/io/druid/query/aggregation/last/DoubleLastAggregatorFactory.java]: changed it to Double, strange that I got both Long and Double wrong

File: processing/src/main/java/io/druid/query/aggregation/first/LongFirstAggregatorFactory.java (hunk 26 above)
Comment (acslk) [inline processing/src/main/java/io/druid/query/aggregation/first/LongFirstAggregatorFactory.java]: changed to separate lines, I agree it looks better

**Comment by gauravkumar37** [comment]
@acslk Thanks for putting in the effort to make this possible. I tried this pull request on the latest master code branch. Though it works for aggregators, I have 2 issues:

- It does not work if the aggregator's output is fed to a having clause. Error:
```
{
  "error": "Unknown exception",
  "errorMessage": "Unknown type[class io.druid.collections.SerializablePair]",
  "errorClass": "com.metamx.common.parsers.ParseException",
  "host": null
}
```
- It may be obvious but the first/last aggregators can only work on the output of the merged/stored data in druid. That is, within a single index-time query granularity, this cannot work and will give random results. I believe this should be highlighted in the docs as well.

**Comment by acslk** [comment]
@gauravkumar37 Thanks for the feedback, here's my thoughts on the issues:
- I haven't run the query using having clause before, but looking at the tests, it seems like aggregators with complex intermediate type such as HyperUnique doesn't work with having clause #2507. I can't think of a nice solution to fix this now, so perhaps we can add similar workaround as #2496 if this feature is needed.
- Maybe I'm not understanding this properly, but is there a use case for this other than ingestion?
- Did you try with "context" : {"finalize" : true}  in inner group by query? It should fix the error, check the comments above for explanation. This is not very intuitive, and should probably go into the doc.

File: processing/src/main/java/io/druid/query/aggregation/JavaScriptAggregatorFactory.java [hunk 27]
```diff
@@ -172,7 +172,7 @@ public AggregatorFactory getMergingFactory(AggregatorFactory other) throws Aggre
               @Override
               public AggregatorFactory apply(String input)
               {
-                return new JavaScriptAggregatorFactory(input, fieldNames, fnAggregate, fnReset, fnCombine, config);
+                return new JavaScriptAggregatorFactory(input, Lists.newArrayList(input), fnCombine, fnReset, fnCombine, config);
```
Comment (jon-wei) [inline processing/src/main/java/io/druid/query/aggregation/JavaScriptAggregatorFactory.java]: Why is this change to JavaScriptAggregatorFactory needed?

Comment (acslk) [inline processing/src/main/java/io/druid/query/aggregation/JavaScriptAggregatorFactory.java]: Previously getCombiningFactory was always called on top of getRequiredColumn to get the identity AggregatorFactory needed for copying the javascript aggregator values. Since getCombiningFactory is no longer called on this, the original getRequiredColumn does really make sense for copying values.

Comment (acslk) [inline processing/src/main/java/io/druid/query/aggregation/JavaScriptAggregatorFactory.java]: Apart from javascript, implementation of getRequiredColumn for other aggregatorFactories seems to work fine without converting to combiningFactory

**Comment by jon-wei** [comment]
LGTM, :+1:

**Comment by drcrallen** [comment]
@acslk if you can fix the conflicts I can help finish up review.

**Comment by gianm** [comment]
since this is one of the last few PRs for 0.9.2, as discussed on the call last week, I'll bump it to 0.9.3.

**Comment by acslk** [comment]
rebased and resolved conflict

**Comment by jon-wei** [comment]
Closing this since @acslk is no longer active on druid development, re-opening at:

https://github.com/druid-io/druid/pull/3566

File: processing/src/main/java/io/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java (hunk 16 above)
Comment (wysstartgo) [reply processing/src/main/java/io/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java]: @Override
 +  public Object deserialize(Object object)
 +  {
 +    Map map = (Map) object;
 +    return new SerializablePair<>(((Number) map.get("lhs")).longValue(), ((Number) map.get("rhs")).doubleValue());
 +  }

why the object is Map type。

please see:https://github.com/druid-io/druid/issues/4734

**Comment by gianm** [comment]
Superseded by #3566.
